src/main/java/net/dv8tion/jda/api/requests/RestRateLimiter.java

```diff
@@ -0,0 +1,243 @@
+    String RESET_HEADER = "X-RateLimit-Reset";
+    /** The number of requests that can be made */
+    String LIMIT_HEADER = "X-RateLimit-Limit";
+    /** The number of remaining requests that can be made */
+    String REMAINING_HEADER = "X-RateLimit-Remaining";
+    /** Returned only on HTTP 429 responses if the rate limit encountered is the global rate limit (not per-route) */
+    String GLOBAL_HEADER = "X-RateLimit-Global";
+    /** A unique string denoting the rate limit being encountered (non-inclusive of top-level resources in the path) */
+    String HASH_HEADER = "X-RateLimit-Bucket";
+    /** The number of seconds to wait before submitting another request */
+    String RETRY_AFTER_HEADER = "Retry-After";
+    /** Returned only on HTTP 429 responses. Value can be user (per bot or user limit), global (per bot or user global limit), or shared (per resource limit) */
+    String SCOPE_HEADER = "X-RateLimit-Scope";
+
+    /**
+     * Initialize your rate limiter with the provided configuration.
+     *
+     * @param config
+     *        The {@link RestConfig} to use
+     */
+    void init(@Nonnull RateLimitConfig config);
+
+    /**
+     * Enqueue a new request.
+     *
+     * <p>Use {@link Work#getRoute()} to determine the correct bucket.
+     *
+     * @param task
+     *        The {@link Work} to enqueue
+     */
+    void enqueue(@Nonnull Work task);
+
+    /**
+     * Indication to stop accepting new requests.
+     *
+     * @param shutdown
+     *        Whether to also cancel previously queued request
+     * @param callback
+     *        Function to call once all requests are completed, used for final cleanup
+     */
+    void stop(boolean shutdown, @Nonnull Runnable callback);
+
+    /**
+     * Whether the queue has stopped accepting new requests.
+     *
+     * @return True, if the queue is stopped
+     */
+    boolean isStopped();
+
+    /**
+     * Cancel all currently queued requests, which are not marked as {@link Work#isPriority() priority}.
+     *
+     * @return The number of cancelled requests
+     */
+    int cancelRequests();
+
+    /**
+     * Type representing a pending request.
+     *
+     * <p>Use {@link #execute()} to run the request (on the calling thread) and {@link #isDone()} to discard it once completed.
+     */
+    interface Work
+    {
+        /**
+         * The {@link Route.CompiledRoute compiled route} of the request.
+         * <br>This is primarily used to handle rate-limit buckets.
+         *
+         * <p>To correctly handle rate-limits, it is recommended to use the {@link #HASH_HEADER bucket hash} header from the response.
+         *
+         * @return The {@link Route.CompiledRoute compiled route}
+         */
+        @Nonnull
+        Route.CompiledRoute getRoute();
+
+        /**
+         * The JDA instance which started the request.
+         *
+         * @return The JDA instance
+         */
+        @Nonnull
+        JDA getJDA();
```

Comment:
Would be an interesting concept to try and consider separating the requester system from relying on the JDA object to better facilitate RestOnly JDA in the future (and allow us to use the requester system in situations where we don't have the obj)

Reply:
Eventually that would be ideal, but that is a non-goal of this PR.